README.md
# NToaster

![Pub Version](https://img.shields.io/pub/v/ntoaster)

https://user-images.githubusercontent.com/36048466/188284228-cb91fd1d-9237-4edc-8d49-c4bba6e7bb31.mp4


#### Notification Toaster.
> Toasts are lightweight notifications designed to mimic the push notifications that have been popularized by mobile and desktop operating systems.
## Features

- Use either a Singleton or an InheritedWidget to get access to the Notifications Center.
- Completely flexible animation and widget UI.
- You can remove the notification programmatically.
- When the user long presses the notification its timer is paused until it is free.

## Usage

1) Attach the notification center manually using the `NToasterController.attach` or Construct
   a `NToasterCenter` which takes care of attaching and detaching the notification center and
   provides you with an `NToaster` `InheritedWidget`.

```dart
// Manual attachment
final ntoaster = NToasterController.getInstance();
ntoaster.attach(context, configuration: NToasterConfiguration
(
)) // Pass your center level configuration;

```

```dart
// Widget attachment
Widget build(BuildContext context) {
  return MaterialApp(
    title: 'Flutter Demo',
    theme: ThemeData(
      primarySwatch: Colors.blue,
    ),
    home: const NToasterCenter(
        child: MyHomePage(title: 'Flutter Demo Home Page')),
  );
}
```

2) Enqueue your notifications (with whichever size/animation)

```dart
// with `InheritedWidget`
final ntoaster = NToaster.of(context);
NotificationMetaData? notification;
notification = NotificationMetaData(
  showFor: Duration(milliseconds:5000),
item: GestureDetector(
  onTap: () {
    notification?.remove();
  },
  child: SizedBox(
    height:100,
    width: 300,
    child: Card(
      elevation: 8.0,
      color: Colors.green,child: Text("$_counter Cool title",style: Theme.of(context).textTheme.subtitle1?.copyWith(color: Colors.white),),),),),
);
ntoaster.enqueue(notification,);
```

3) If you hold into a reference of a `NotificationMetaData` you can `dequeue` it immediately before
   the timer ends.

```dart

final notificationX = NotificationMetaData(...);
final ntoaster = NToaster.of(context);
ntoaster.dequeue(notificationX);
```



CHANGELOG.md
## 1.0.0

* docs: Add note for NToasterCenter usage
* fix: Properly clean up after detaching
* perf: Optimize lookups for timers
* chore: Bump to 1.0.0 (major release)

## 0.0.1

* Initial Release.



lib/src/ntoaster.dart
import 'package:flutter/widgets.dart';
import 'package:ntoaster/ntoaster.dart';
import 'package:ntoaster/src/ntoaster_base.dart';

/// Should wrap a widget that is not subject to removal from the Widget tree.
class NToasterCenter extends StatefulWidget {
  const NToasterCenter(
      {Key? key,
      required this.child,
      this.configuration = const NToasterConfiguration()})
      : super(key: key);
  final NToasterConfiguration configuration;
  final Widget child;

  @override
  State<NToasterCenter> createState() => _NToasterCenterState();
}

class _NToasterCenterState extends State<NToasterCenter> {
  final NToasterController _controller = NToasterController.getInstance();

  @override
  void initState() {
    super.initState();
    _controller.attach(context, configuration: widget.configuration);
  }

  @override
  Widget build(BuildContext context) {
    return NToaster(
      _controller,
      child: widget.child,
    );
  }

  @override
  void dispose() {
    super.dispose();
    _controller.detach();
  }
}

class NToaster extends InheritedWidget implements NToasterBase {
  const NToaster(
    this._controller, {
    Key? key,
    required Widget child,
  }) : super(key: key, child: child);
  final NToasterController _controller;

  static NToaster of(BuildContext context) {
    final NToaster? result =
        context.dependOnInheritedWidgetOfExactType<NToaster>();
    assert(result != null, 'No NToaster found in context');
    return result!;
  }

  @override
  bool updateShouldNotify(NToaster oldWidget) => false;

  @override
  void attach(BuildContext context,
          {NToasterConfiguration configuration =
              const NToasterConfiguration()}) =>
      _controller.attach(context, configuration: configuration);

  @override
  void dequeue(NotificationMetaData data) => _controller.dequeue(data);

  @override
  void detach() => _controller.detach();

  @override
  void enqueue(NotificationMetaData data) => _controller.enqueue(data);
}


lib/src/ntoaster_controller.dart
import 'dart:collection';
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:ntoaster/src/configuration/notification_meta_data.dart';
import 'package:ntoaster/src/configuration/ntoaster_configuration.dart';
import 'package:ntoaster/src/ntoaster_base.dart';
import 'package:pausable_timer/pausable_timer.dart';

class NToasterController implements NToasterBase {
  bool get isAttached => _listWasAttached;

  static NToasterController? _instance;
  final List<NotificationMetaData> _queue = <NotificationMetaData>[];

  final LinkedHashMap<int, PausableTimer> _timers = LinkedHashMap();

  NToasterConfiguration? _configuration;

  final _key = GlobalKey<AnimatedListState>();
  OverlayEntry? _entry;
  var _listWasAttached = false;

  NToasterController._();

  static NToasterController getInstance() {
    if (_instance != null) return _instance!;
    _instance = NToasterController._();
    return _instance!;
  }

  /// Attaches the notification center with [configuration]
  @override
  void attach(BuildContext context,
      {NToasterConfiguration configuration = const NToasterConfiguration()}) {
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      if (!_listWasAttached) {
        _configuration = configuration;
        final entry = OverlayEntry(
            maintainState: true,
            builder: (context) => PositionedDirectional(
                top: _configuration?.top,
                bottom: _configuration?.bottom,
                start: _configuration?.start,
                end: _configuration?.end,
                child: Align(
                  alignment: AlignmentDirectional.topEnd,
                  child: Container(
                    constraints: BoxConstraints(
                      maxWidth: _configuration?.maxWidth ??
                          MediaQuery.of(context).size.width * 0.3,
                    ),
                    child: Material(
                      color: Colors.transparent,
                      child: AnimatedList(
                          key: _key,
                          initialItemCount: _queue.length,
                          shrinkWrap: true,
                          itemBuilder: (context, index, animation) {
                            final item = _queue.elementAt(index);
                            return Center(
                                key: ValueKey(item.id),
                                child: GestureDetector(
                                    behavior: HitTestBehavior.translucent,
                                    onLongPress: () {
                                      final timer = _timers.entries
                                          .elementAt(index)
                                          .value;
                                      timer.pause();
                                    },
                                    onLongPressUp: () {
                                      final timer = _timers.entries
                                          .elementAt(index)
                                          .value;
                                      timer.start();
                                    },
                                    child: Padding(
                                      padding: _configuration?.contentPadding ??
                                          const EdgeInsets.only(bottom: 8.0),
                                      child: item.insertionAnimationBuilder
                                              ?.call(context, item.item,
                                                  animation) ??
                                          item.item,
                                    )));
                          }),
                    ),
                  ),
                )));
        final overlay = Overlay.of(context);
        if (overlay != null) {
          _entry = entry;
          overlay.insert(entry);
          _listWasAttached = true;
        } else {
          throw FlutterError(
              "You can't attach the notifications center when there is not an Overlay widget above it.");
        }
      }
    });
  }

  @override
  void enqueue(NotificationMetaData data) {
    if (_listWasAttached) {
      final length = _queue.length;
      _queue.add(data);
      _key.currentState?.insertItem(length);
      final timer = PausableTimer(data.showFor, () {
        try {
          _removeItem(data);
        } catch (e) {
          log(e.toString());
        }
      });
      _timers[data.id] = timer;
      timer.start();
    } else {
      throw FlutterError(
          "You can't enqueue notifications before attaching the notification center with [NToast.attach]");
    }
  }

  @override
  void dequeue(NotificationMetaData data) {
    if (_listWasAttached && _queue.isNotEmpty) {
      _removeItem(data);
    } else {
      throw FlutterError(
          "You can't dequeue notifications before attaching the notification center with [NToast.attach] or when the notification center is empty.");
    }
  }

  void _removeItem(NotificationMetaData data) {
    final indexOfItem = _queue.indexWhere((element) => element.id == data.id);
    if (indexOfItem != -1) {
      final item = _queue[indexOfItem];
      _key.currentState?.removeItem(
          indexOfItem,
          (context, animation) => Center(
                child: Padding(
                  padding: _configuration?.contentPadding ??
                      const EdgeInsets.only(bottom: 8),
                  child: item.removalAnimationBuilder
                          ?.call(context, item.item, animation) ??
                      item.item,
                ),
              ));
      _queue.removeAt(indexOfItem);
      _timers.remove(data.id);
    }
  }

  /// Detaches the notification center with [configuration]
  @override
  void detach() {
    _entry?.remove();
    _listWasAttached = false;
    _queue.clear();
    _timers.clear();
    _configuration = null;
  }
}


lib/src/configuration/notification_meta_data.dart
import 'dart:math';

import 'package:flutter/widgets.dart';
import 'package:ntoaster/ntoaster.dart';

class NotificationMetaData {
  late int id;

  /// The widget holding your notification/toast configuration
  final Widget item;

  /// How much time you want the notification to be shown for
  final Duration showFor;

  /// Your [item] enter transition
  ///
  /// Use [animation] to drive your own animation like this:
  ///
  /// ```dart
  /// Tween<Offset>(
  ///   begin: Offset.zero,
  ///   end: Offset(1, 0)
  /// ).animate(animation)
  /// ```
  /// Or use it directly.
  final Widget Function(
      BuildContext context,

      /// The passed [item]
      Widget? child,
      Animation<double> animation)? insertionAnimationBuilder;

  /// Your [item] exit transition
  ///
  /// Use [animation] to drive your own animation like this:
  ///
  /// ```dart
  /// Tween<Offset>(
  ///   begin: Offset.zero,
  ///   end: Offset(1, 0)
  /// ).animate(animation)
  /// ```
  /// Or use it directly.
  final Widget Function(
      BuildContext context,

      /// The passed [item]
      Widget? child,
      Animation<double> animation)? removalAnimationBuilder;

  NotificationMetaData({
    required this.item,
    this.showFor = const Duration(milliseconds: 2000),
    this.removalAnimationBuilder,
    this.insertionAnimationBuilder,
  }) {
    id = Random(DateTime.now().millisecondsSinceEpoch).nextInt(10000000);
  }

  /// Removes this notification (triggering the [removalAnimationBuilder])
  void remove() {
    final ntoaster = NToasterController.getInstance();

    ntoaster.dequeue(this);
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:ntoaster/ntoaster.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const NToasterCenter(
          child: MyHomePage(title: 'Flutter Demo Home Page')),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  void initState() {
    super.initState();
  }

  int _counter = 0;

  void _incrementCounter() {
    _counter++;
    setState(() {});
    final ntoaster = NToaster.of(context);
    NotificationMetaData? notification;
    notification = NotificationMetaData(
      insertionAnimationBuilder: (context, child, animation) => ScaleTransition(
        scale: animation,
        child: child,
      ),
      removalAnimationBuilder: (context, child, animation) => SlideTransition(
        position: Tween(begin: const Offset(1, 0), end: const Offset(0.5, 0))
            .animate(animation),
        child: child,
      ),
      showFor: Duration(
          milliseconds:
              (Random(DateTime.now().millisecondsSinceEpoch).nextDouble() *
                      5000)
                  .toInt()),
      item: GestureDetector(
        onTap: () {
          notification?.remove();
        },
        child: SizedBox(
          height:
              Random(DateTime.now().millisecondsSinceEpoch).nextDouble() * 100 +
                  50,
          width: 300,
          child: Card(
            elevation: 8.0,
            color: Colors.red.withGreen(
                Random(DateTime.now().millisecondsSinceEpoch).nextInt(255)),
            child: Center(
              child: Text(
                "$_counter Cool title",
                style: Theme.of(context)
                    .textTheme
                    .subtitle1
                    ?.copyWith(color: Colors.white),
              ),
            ),
          ),
        ),
      ),
    );
    ntoaster.enqueue(
      notification,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}
